### ISSUE TITLE
Implementation of ray/ray intersections

### ISSUE THREAD

**Issue by You-Ay** [You-Ay/vectors#4]
Two rays g and h may be parallel, skew, or intersecting. If they intersect, they have a point of intersection P corresponding to values r, s of the ray parameters. This is to be evaluated by the routine 
```C
intersection_pointlike intersect_ray_ray(const ray *g, const ray *h)
```
All relevant information can then be stored in the `intersection_pointlike` return value (e.g. for later print to screen).

**Comment by You-Ay** [comment]
Solving for the intersection comes down to solving a system of three linear equations in the variables r and s, the two ray parameters. I suggest to first consider two of these equations and to solve them directly (without steps) by the explicit formula that comes from Cramer's rule, see 

https://en.m.wikipedia.org/wiki/Cramer%27s_rule#Explicit_formulas_for_small_systems

The final equation is then used to check consistency (intersecting vs skew rays).

**Comment by You-Ay** [comment]
For those who don't like to teach C how to solve systems of linear equations... I just imagined an alternative way of calculating the intersection of two rays:

_g : x = u<sub>g</sub> + r v<sub>g</sub>_
_h : x = u<sub>h</sub> + s v<sub>h</sub>_

First, we have to assure that _g_ and _h_ are neither identical nor parallel. Then, we are sure to get a non-zero cross product:

_n = v<sub>g</sub> X v<sub>h</sub>_

We now promote _h_ to a plane by adding _n_ as a second directional vector:

_E : x = u<sub>h</sub> + s v<sub>h</sub> + t n_

Rather than calculating the intersection between _g_ and _h_, we calculate the intersection between _g_ and _E_! We are sure that such an intersection exists.

The program does already convert the plane to cartesian form. Plugging in _g_ into this cartesian form gives a single equation in one variable _r_. This is easily solved.

Finally, we have to check whether the intersection point between _g_ and _E_ lies also on _h_. If so, the rays are intersecting, if not, they are skew.

**Comment by Const-He** [comment]
It might be a good idea to add a function that compares vectors and checks for equality in order to simplify the intersection calculations. (eg check if two vectors are colinear by comparing their unit vectors)

**Comment by You-Ay** [comment]
Great idea to encapsulate this in functions. Makes the code more transparent, modular, and readable. According to your suggestion, I've added three check functions `bool are_equal(vector v, vector w)`, `bool are_collinear(...)`, `bool are_orthogonal` to vector.h.

There is a tricky thing, however. In our ray tracing, we will have lots of computations, and rounding errors can quickly add up to become sizeable. Checks should thus not be done for real equality between two doubles, but rather against some reasonable numerical limit. That's why I introduced the global constant variable `const double epsilon`. 10^-6 appeares a possible value if we assume reasonable scales. I've given an example for a check function in `vector.c`
